blocks/log/sqlite.py

```diff
@@ -0,0 +1,240 @@
+    else:
+        return adapt_obj(obj)
+
+
+def _get_row(row, key):
+    """Handle the returned row e.g. unpickle if needed."""
+    if row is not None:
+        value = row[0]
+        # Resumption UUIDs are stored as bytes and should not be unpickled
+        if (isinstance(value, (sqlite3.Binary, bytes)) and
+                key != 'resumed_from'):
+            value = cPickle.loads(bytes(value))
+        return value
+    raise KeyError(key)
+
+
+def _register_adapter(value, key):
+    """Register an adapter if the type of value is unknown."""
+    # Assuming no storage of non-simple types on channel 'resumed_from'
+    if (not isinstance(value, (type(None), int, float, six.string_types,
+                               bytes, numpy.ndarray)) and
+            key != 'resumed_from'):
+        sqlite3.register_adapter(type(value), adapt_obj)
+
+
+class SQLiteLog(_TrainingLog, Mapping):
+    r"""Training log using SQLite as a backend.
+
+    Parameters
+    ----------
+    database : str, optional
+        The database (file) to connect to. Can also be `:memory:`. See
+        :func:`sqlite3.connect` for details. Uses `config.sqlite_database`
+        by default.
+    \*\*kwargs
+        Arguments to pass to :class:`_TrainingLog`
+
+    Notes
+    -----
+    .. todo::
+
+       Currently this log ignores previous logs in case of resumption.
+
+    """
+    def __init__(self, database=None, **kwargs):
+        if database is None:
+            database = config.sqlite_database
+        self.database = database
+        self.conn = sqlite3.connect(database)
+        sqlite3.register_adapter(numpy.ndarray, adapt_ndarray)
+        with self.conn:
+            self.conn.execute("""CREATE TABLE IF NOT EXISTS entries (
+                                   uuid BLOB NOT NULL,
+                                   time INT NOT NULL,
+                                   "key" TEXT NOT NULL,
+                                   value,
+                                   PRIMARY KEY(uuid, time, "key")
+                                 );""")
+            self.conn.execute("""CREATE TABLE IF NOT EXISTS status (
+                                   uuid BLOB NOT NULL,
+                                   "key" text NOT NULL,
+                                   value,
+                                   PRIMARY KEY(uuid, "key")
+                                 );""")
+        self.status = SQLiteStatus(self)
+        super(SQLiteLog, self).__init__(**kwargs)
+
+    def __getstate__(self):
+        """Retrieve the state for pickling.
+
+        :class:`sqlite3.Connection` objects are not picklable, so the
+        `conn` attribute is removed and the connection re-opened upon
+        unpickling.
+
+        """
+        state = self.__dict__.copy()
+        del state['conn']
+        self.resume()
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self.conn = sqlite3.connect(self.database)
```

Comment:
Will this cause the same sorts of problems that we did gymnastics to avoid with datasets? i.e. should we make this lazy, so that you can unpickle even if your database attribute is somewhere else (because you moved machines, for example, or because you're trying to use someone else's checkpoint, or whatever)?